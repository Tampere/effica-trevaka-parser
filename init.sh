#!/bin/bash

# SPDX-FileCopyrightText: 2023-2024 Tampere region
#
# SPDX-License-Identifier: LGPL-2.1-or-later

set -e
set -u

function create_database() {
  psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" <<-EOSQL
    CREATE DATABASE $1;
    \connect $1
    CREATE SCHEMA migration;
EOSQL
}

create_database vesilahti
