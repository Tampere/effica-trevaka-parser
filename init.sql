-- SPDX-FileCopyrightText: 2021 City of Tampere
--
-- SPDX-License-Identifier: LGPL-2.1-or-later

-- Add extensions here as needed (for UUIDs at least)
CREATE SCHEMA IF NOT EXISTS ext;
CREATE SCHEMA IF NOT EXISTS migration;

CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA ext;
CREATE EXTENSION IF NOT EXISTS btree_gist WITH SCHEMA ext;