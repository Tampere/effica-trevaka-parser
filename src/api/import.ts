// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import express from "express"
import { config } from "../config"
import { executePostImportFixes, importFileData } from "../import/service"
import { readFilesFromDir } from "../io/io"
import { postImportFixes } from "../mapping/citySpecific"
import { FileDescriptor, ImportOptions } from "../types"
import { ErrorWithCause } from "../util/error"
import { time, timeEnd } from "../util/timing"

const router = express.Router();
router.get("/", async (req, res, next) => {
    time("**** Import total ", undefined, "*")
    const reqPath = req.query.path ?? "/xml"
    const basePath = `${__dirname}/../..`
    const path = basePath + reqPath
    const importOptions: ImportOptions = { path, returnAll: req.query.returnAll === "true" }
    try {
        const files: FileDescriptor[] = await readFilesFromDir(path)
        const importResult = await importFileData(files, importOptions)
        res.status(200).json(importResult)
    } catch (err) {
        next(new ErrorWithCause(`Import operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Import total ", undefined, "*")
})

router.get("/fix", async (req, res, next) => {
    const cityFixScripts = postImportFixes[config.cityVariant] ?? []

    try {
        if (cityFixScripts.length > 0) {
            await executePostImportFixes(cityFixScripts)
            res.status(200).json(`Import fixes for city variant '${config.cityVariant}' successfully run`)
        } else {
            res.status(200).json(`No import fixes found for city variant: '${config.cityVariant}'`)
        }
    } catch (err) {
        next(new ErrorWithCause(`Import fixes for city variant '${config.cityVariant}' failed, transaction rolled back:`, err))
    }
})

export default router