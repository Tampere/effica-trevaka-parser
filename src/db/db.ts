import pgPromise from "pg-promise"
import { config } from "../config"

export const pgp = pgPromise({})

// date parsing seems to be scuffed, does local tz transformation but declares it UTC -> corrupts dates
//pgp.pg.types.setTypeParser(1082, dateParser)

const migrationDb = pgp({ ...config.migrationDb })

export default migrationDb