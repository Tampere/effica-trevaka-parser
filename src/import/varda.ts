// SPDX-FileCopyrightText: 2021-2022 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import { config } from "../config";
import migrationDb, { pgp } from "../db/db";
import { baseQueryParameters } from "../util/queryTools";
import {
    VardaClient,
    VardaV1Children,
    VardaV1Person,
} from "../util/varda-client";

export const importVarda = async (vardaClient: VardaClient) => {
    await migrationDb.none(tableSql, baseQueryParameters);
    await importVardaChildData(vardaClient);
    return await importVardaPersonData(vardaClient);
};

const importVardaChildData = async (vardaClient: VardaClient) => {
    let children = await vardaClient.getChildren();
    const childrenCount = children.count;
    const pageCount = Math.ceil(childrenCount / children.results.length);
    console.log(`Children count=${childrenCount}, page count=${pageCount}`);
    let pageNumber = 1;
    while (true) {
        const insertSql = pgp.helpers.insert(children.results, childColumns);
        await migrationDb.none(insertSql);

        const nextUrl = children.next;
        if (nextUrl === null) {
            break;
        }

        if (++pageNumber % 50 === 0) {
            console.log(`Loading page ${pageNumber}`);
        }
        const data = await vardaClient.getByUrl<VardaV1Children>(nextUrl);
        if (data === null) {
            throw Error(`Next url '${nextUrl}' returned 404`);
        }
        children = data;
    }
};

export const importVardaPersonData = async (vardaClient: VardaClient) => {
    const data: Array<VardaChildRow> = [];

    const children = await migrationDb.manyOrNone<VardaChildRow>(
        selectSql,
        baseQueryParameters
    );
    let rowNumber = 0;
    for (const child of children) {
        if (rowNumber++ % 500 === 0) {
            console.log(`Loading person with row number ${rowNumber}`);
        }
        const { id: childId, henkilo: personUrl } = child;
        const person = await vardaClient.getByUrl<VardaV1Person>(personUrl);
        if (person === null) {
            console.warn(`Person url '${personUrl}' returned 404`);
            continue;
        }

        const updateSql =
            pgp.helpers.update(person, personColumns) +
            pgp.as.format(" WHERE id = ${id} RETURNING *", { id: childId });
        data.push(await migrationDb.one<VardaChildRow>(updateSql));
    }

    return data;
};

interface VardaChildRow {
    id: number;
    henkilo: string;
    henkilo_oid: string;
    vakatoimija_oid: string | null;
    paos_organisaatio_oid: string | null;
    paos_kytkin: boolean;
    henkilo_id: number;
    etunimet: string | null;
    kutsumanimi: string | null;
    sukunimi: string | null;
    syntyma_pvm: Date | null;
}

const tableSql = `
    CREATE TABLE IF NOT EXISTS $(migrationSchema:name).varda_child (
        id BIGINT PRIMARY KEY,
        henkilo TEXT NOT NULL,
        henkilo_oid TEXT NOT NULL,
        vakatoimija_oid TEXT,
        paos_organisaatio_oid TEXT,
        paos_kytkin BOOLEAN NOT NULL,
        henkilo_id BIGINT,
        etunimet TEXT,
        kutsumanimi TEXT,
        sukunimi TEXT,
        syntyma_pvm DATE,
        CHECK (vakatoimija_oid IS NOT NULL OR paos_organisaatio_oid IS NOT NULL)
    );
`;

const selectSql = `
    SELECT *
    FROM $(migrationSchema:name).varda_child
    WHERE henkilo_id IS NULL
    ORDER BY id;
`;

const childColumns = new pgp.helpers.ColumnSet(
    [
        "id",
        "henkilo",
        "henkilo_oid",
        "vakatoimija_oid",
        "paos_organisaatio_oid",
        "paos_kytkin",
    ],
    {
        table: {
            table: "varda_child",
            schema: config.migrationSchema,
        },
    }
);

const personColumns = new pgp.helpers.ColumnSet(
    [
        {
            name: "henkilo_id",
            prop: "id",
        },
        "etunimet",
        "kutsumanimi",
        "sukunimi",
        "syntyma_pvm",
    ],
    {
        table: {
            table: "varda_child",
            schema: config.migrationSchema,
        },
    }
);
