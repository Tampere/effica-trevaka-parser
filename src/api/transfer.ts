// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import express from "express"
import { transferAbsences } from "../transfer/absence"
import { transferApplicationData } from "../transfer/application"
import { transferAssistanceActionsData } from "../transfer/assistance-actions"
import { transferAssistanceNeedsData } from "../transfer/assistance-needs"
import { transferBackupCares } from "../transfer/backup-cares"
import { transferChildAttendances } from "../transfer/child-attendances"
import { transferDaycareData } from "../transfer/daycare"
import { transferDaycareOidData } from "../transfer/daycare-oid"
import { transferDepartmentData } from "../transfer/departments"
import { transferUnitManagerData } from "../transfer/evaka-unit-manager"
import { transferFamiliesData } from "../transfer/families"
import { transferFeeAlterationsData } from "../transfer/fee-alterations"
import { transferFeeDecisionData } from "../transfer/fee-decisions"
import { transferIncomeData } from "../transfer/income"
import { transferPersonData } from "../transfer/person"
import { transferPlacementsData } from "../transfer/placements"
import { transferVarda } from "../transfer/varda"
import { transferVoucherValueDecisions } from "../transfer/voucher-value-decisions"
import { MigrationOperation } from "../types/internal"
import { ErrorWithCause } from "../util/error"
import { time, timeEnd } from "../util/timing"

const dependencyOrder: MigrationOperation[] =
    [
        { name: "persons", function: transferPersonData },
        { name: "families", function: transferFamiliesData },
        { name: "assistance_needs", function: transferAssistanceNeedsData },
        { name: "assistance_actions", function: transferAssistanceActionsData },
        { name: "income", function: transferIncomeData },
        // daycare transfers suspended after locking down daycare data in production
        //{ name: "unit_manager", function: transferUnitManagerData },
        //{ name: "daycare", function: transferDaycareData },
        { name: "departments", function: transferDepartmentData },
        { name: "placements", function: transferPlacementsData },
        { name: "fee_alterations", function: transferFeeAlterationsData },
        { name: "voucher_value_decisions", function: transferVoucherValueDecisions },
        { name: "fee_decisions", function: transferFeeDecisionData },
        { name: "application", function: transferApplicationData },
        { name: "absences", function: transferAbsences },
        { name: "backup_care", function: transferBackupCares },
        { name: "child_attendances", function: transferChildAttendances },
        { name: "daycare_oid", function: transferDaycareOidData }
    ]


const router = express.Router()
router.get("/daycare", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer daycare total ", undefined, "*")
    try {
        const results = await transferDaycareData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer daycare total ", undefined, "*")
})
router.get("/departments", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer departments total ", undefined, "*")
    try {
        const results = await transferDepartmentData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer departments total ", undefined, "*")
})
router.get("/persons", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer persons total ", undefined, "*")
    try {
        const results = await transferPersonData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer persons total ", undefined, "*")
})
router.get("/families", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer families total ", undefined, "*")
    try {
        const results = await transferFamiliesData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer families total ", undefined, "*")
})
router.get("/assistance_needs", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer assistance needs total ", undefined, "*")
    try {
        const results = await transferAssistanceNeedsData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer assistance needs total ", undefined, "*")
})
router.get("/assistance_actions", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer assistance actions total ", undefined, "*")
    try {
        const results = await transferAssistanceActionsData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer assistance actions total ", undefined, "*")
})
router.get("/application", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer application total ", undefined, "*")
    try {
        const results = await transferApplicationData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer application total ", undefined, "*")
})
router.get("/placements", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer families total ", undefined, "*")
    try {
        const results = await transferPlacementsData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer families total ", undefined, "*")
})
router.get("/fee_alterations", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer fee alterations total ", undefined, "*")
    try {
        const results = await transferFeeAlterationsData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer fee alterations total ", undefined, "*")
})
router.get("/voucher_value_decisions", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer voucher value decisions total ", undefined, "*")
    try {
        const results = await transferVoucherValueDecisions(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer voucher value decisions total ", undefined, "*")
})
router.get("/fee_decisions", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer fee decisions total ", undefined, "*")
    try {
        const results = await transferFeeDecisionData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer fee decisions total ", undefined, "*")
})
router.get("/absences", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer absences total ", undefined, "*")
    try {
        const results = await transferAbsences(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer absences total ", undefined, "*")
})
router.get("/backup_cares", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer backup cares total ", undefined, "*")
    try {
        const results = await transferBackupCares(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer backup cares total ", undefined, "*")
})
router.get("/child_attendances", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer child attendances total ", undefined, "*")
    try {
        const results = await transferChildAttendances(returnAll)
        res.status(200).json(results)
    } catch (err) {
        console.log(err)
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer child attendances total ", undefined, "*")
})
router.get("/unit_manager", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer unit managers total ", undefined, "*")
    try {
        const results = await transferUnitManagerData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer unit managers total ", undefined, "*")
})

router.get("/income", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer income total ", undefined, "*")
    try {
        const results = await transferIncomeData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer income total ", undefined, "*")
})

router.get("/daycare_oid", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer daycare varda ids total ", undefined, "*")
    try {
        const results = await transferDaycareOidData(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer daycare varda ids total ", undefined, "*")
})

router.get("/varda", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer varda total ", undefined, "*")
    try {
        const results = await transferVarda(returnAll)
        res.status(200).json(results)
    } catch (err) {
        next(new ErrorWithCause(`Transfer operation failed, transaction rolled back:`, err))
    }
    timeEnd("**** Transfer varda total ", undefined, "*")
})


router.get("/", async (req, res, next) => {
    const returnAll = req.query.returnAll === "true"
    time("**** Transfer all total ", undefined, "*")
    const results: Record<string, any> = {}
    let status = 200
    for (let operation of dependencyOrder) {
        try {
            results[operation.name] = await operation.function(returnAll)
        } catch (error) {
            results[operation.name] = error
            status = 500
            break
        }
    }

    res.status(status).json(results)

    timeEnd("**** Transfer all total ", undefined, "*")
})

export default router