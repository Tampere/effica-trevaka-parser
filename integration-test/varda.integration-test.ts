// SPDX-FileCopyrightText: 2021-2022 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import db from "../src/db/db";
import { importVarda, importVardaUnitData } from "../src/import/varda";
import { initDb } from "../src/init";
import { transferVarda } from "../src/transfer/varda";
import { transformVarda } from "../src/transform/varda";
import {
    cleanupDb,
    setupTable,
    setupTransfer,
    setupTransformation
} from "../src/util/testTools";
import {
    VardaClient,
    VardaV1Child,
    VardaV1Page,
    VardaV1Person,
    VardaV1Unit
} from "../src/util/varda-client";
import migrationDb from "../src/db/db";

beforeAll(async () => {
    await migrationDb.tx(async (tx) => await initDb(tx));
});

beforeEach(async () => {
    await db.tx(tx => cleanupDb(tx));
    await setupTable("persons");
    await setupTable("codes");
    await setupTransformation("persons");
    await setupTransfer("persons");
});

afterAll(async () => {
    await db.$pool.end();
});

it("should import varda unit data", async () => {
    const mockDataByUrl: Record<string, VardaV1Page<VardaV1Unit>> = {
        "localhost/v1/toimipaikat": {
            count: 3,
            next: "localhost/v1/toimipaikat?page=2",
            previous: null,
            results: [
                {
                    url: "localhost/toimipaikka/11",
                    lahdejarjestelma: "93",
                    id: 11,
                    vakajarjestaja: "localhost/organisaatio/2.2.2.1",
                    vakajarjestaja_oid: "2.2.2.1",
                    organisaatio_oid: "",
                    nimi: "Toimipaikka 1",
                    alkamis_pvm: "2022-01-01",
                    paattymis_pvm: null,
                    hallinnointijarjestelma: "VARDA",
                    tunniste: "tunniste1",
                    muutos_pvm: "2022-02-09T11:29:00.249Z",
                },
                {
                    url: "localhost/toimipaikka/12",
                    lahdejarjestelma: "45",
                    id: 12,
                    vakajarjestaja: "localhost/organisaatio/2.2.2.4",
                    vakajarjestaja_oid: "2.2.2.4",
                    organisaatio_oid: "",
                    nimi: "Toimipaikka 2",
                    alkamis_pvm: "2020-01-01",
                    paattymis_pvm: "2020-12-31",
                    hallinnointijarjestelma: "VARDA",
                    tunniste: "tunniste2",
                    muutos_pvm: "2022-02-09T11:29:00.249Z",
                },
            ],
        },
        "localhost/v1/toimipaikat?page=2": {
            count: 3,
            next: null,
            previous: "localhost/v1/toimipaikat",
            results: [
                {
                    url: "localhost/toimipaikka/13",
                    lahdejarjestelma: "57",
                    id: 13,
                    vakajarjestaja: "localhost/organisaatio/2.2.2.3",
                    vakajarjestaja_oid: "2.2.2.3",
                    organisaatio_oid: "",
                    nimi: "Toimipaikka 3",
                    alkamis_pvm: "2021-01-01",
                    paattymis_pvm: null,
                    hallinnointijarjestelma: "VARDA",
                    tunniste: "tunniste3",
                    muutos_pvm: "2022-02-09T11:29:00.249Z",
                },
            ],
        },
    };
    const client: VardaClient = {
        //@ts-ignore
        getUnits: jest.fn(() => mockDataByUrl["localhost/v1/toimipaikat"]),
        //@ts-ignore
        getByUrl: jest.fn((url) => {
            const mockData = mockDataByUrl[url];
            if (mockData === undefined) {
                return Promise.reject("404");
            }
            return Promise.resolve(mockData);
        }),
    };

    const importData = await importVardaUnitData(client);
    expect(importData).toMatchSnapshot();
});

it("should import, transform and transfer varda data", async () => {
    const mockDataByUrl: Record<
        string,
        VardaV1Page<VardaV1Child> | VardaV1Person
    > = {
        "localhost//v1/lapset": {
            count: 3,
            next: "localhost//v1/lapset?page=2",
            previous: null,
            results: [
                {
                    url: "localhost/lapsi/11",
                    lahdejarjestelma: "93",
                    id: 11,
                    henkilo: "localhost/henkilo/21",
                    henkilo_oid: "1.1.1.1",
                    vakatoimija: "localhost/organisaatio/2.2.2.1",
                    vakatoimija_oid: "2.2.2.1",
                    oma_organisaatio: null,
                    oma_organisaatio_oid: null,
                    paos_organisaatio: null,
                    paos_organisaatio_oid: null,
                    paos_kytkin: false,
                    varhaiskasvatuspaatokset_top: [
                        "localhost/paatos/1",
                        "localhost/paatos/2",
                    ],
                    tunniste: "tunniste1",
                    muutos_pvm: "2022-02-09T11:29:00.249Z",
                },
                {
                    url: "localhost/lapsi/12",
                    lahdejarjestelma: "45",
                    id: 12,
                    henkilo: "localhost/henkilo/22",
                    henkilo_oid: "1.1.1.2",
                    vakatoimija: null,
                    vakatoimija_oid: null,
                    oma_organisaatio_nimi: "Kaupunki 1",
                    oma_organisaatio: "localhost/organisaatio/2.2.2.4",
                    oma_organisaatio_oid: "2.2.2.4",
                    paos_organisaatio_nimi: "Perhepäivähoitaja 1",
                    paos_organisaatio: "localhost/organisaatio/2.2.2.2",
                    paos_organisaatio_oid: "2.2.2.2",
                    paos_kytkin: true,
                    varhaiskasvatuspaatokset_top: ["localhost/paatos/3"],
                    tunniste: null,
                    muutos_pvm: "2022-02-09T11:29:00.249Z",
                },
            ],
        },
        "localhost//v1/lapset?page=2": {
            count: 3,
            next: null,
            previous: "localhost//v1/lapset",
            results: [
                {
                    url: "localhost/lapsi/13",
                    lahdejarjestelma: "57",
                    id: 13,
                    henkilo: "localhost/henkilo/21",
                    henkilo_oid: "1.1.1.1",
                    vakatoimija: "localhost/organisaatio/2.2.2.3",
                    vakatoimija_oid: "2.2.2.3",
                    oma_organisaatio: null,
                    oma_organisaatio_oid: null,
                    paos_organisaatio: null,
                    paos_organisaatio_oid: null,
                    paos_kytkin: false,
                    varhaiskasvatuspaatokset_top: [],
                    tunniste: "tunniste2",
                    muutos_pvm: "2022-02-09T11:29:47.317Z",
                },
            ],
        },
        "localhost/henkilo/21": {
            url: "localhost/henkilo/21",
            id: 21,
            etunimet: "Christian",
            kutsumanimi: "Christian",
            sukunimi: "Child",
            henkilo_oid: "1.1.1.1",
            syntyma_pvm: "2019-06-01",
            lapsi: ["localhost/lapsi/11", "localhost/lapsi/13"],
            tyontekija: [],
            turvakielto: false,
        },
        "localhost/henkilo/22": {
            url: "localhost/henkilo/22",
            id: 22,
            etunimet: "Camilla",
            kutsumanimi: "Camilla",
            sukunimi: "Child",
            henkilo_oid: "1.1.1.2",
            syntyma_pvm: "2018-11-25",
            lapsi: ["localhost/lapsi/12"],
            tyontekija: ["localhost/tyontekija/1"],
            turvakielto: true,
        },
    };
    const client: VardaClient = {
        //@ts-ignore
        getChildren: jest.fn(() => mockDataByUrl["localhost//v1/lapset"]),
        //@ts-ignore
        getByUrl: jest.fn((url) => {
            const mockData = mockDataByUrl[url];
            if (mockData === undefined) {
                return Promise.reject("404");
            }
            return Promise.resolve(mockData);
        }),
    };

    const importData = await importVarda(client);
    expect(importData).toMatchSnapshot();

    const transformData = await transformVarda(true);
    expect(transformData).toMatchSnapshot(Array(3).fill(transformExpectation));

    const transferData = await transferVarda(true);
    expect(transferData).toMatchSnapshot(Array(3).fill(transferExpectation));
});

const transformExpectation = {
    evaka_person_id: expect.any(String),
    row_id: expect.any(String),
};

const transferExpectation = {
    created_at: expect.any(Date),
    modified_at: expect.any(Date),
    evaka_person_id: expect.any(String),
};
