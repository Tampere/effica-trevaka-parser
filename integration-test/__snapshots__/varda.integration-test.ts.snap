// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should import varda unit data 1`] = `
Array [
  Object {
    "alkamis_pvm": 2022-01-01T00:00:00.000Z,
    "hallinnointijarjestelma": "VARDA",
    "id": "11",
    "lahdejarjestelma": "93",
    "muutos_pvm": "2022-02-09T11:29:00.249Z",
    "nimi": "Toimipaikka 1",
    "organisaatio_oid": "",
    "paattymis_pvm": null,
    "tunniste": "tunniste1",
    "url": "localhost/toimipaikka/11",
    "vakajarjestaja": "localhost/organisaatio/2.2.2.1",
    "vakajarjestaja_oid": "2.2.2.1",
  },
  Object {
    "alkamis_pvm": 2020-01-01T00:00:00.000Z,
    "hallinnointijarjestelma": "VARDA",
    "id": "12",
    "lahdejarjestelma": "45",
    "muutos_pvm": "2022-02-09T11:29:00.249Z",
    "nimi": "Toimipaikka 2",
    "organisaatio_oid": "",
    "paattymis_pvm": 2020-12-31T00:00:00.000Z,
    "tunniste": "tunniste2",
    "url": "localhost/toimipaikka/12",
    "vakajarjestaja": "localhost/organisaatio/2.2.2.4",
    "vakajarjestaja_oid": "2.2.2.4",
  },
  Object {
    "alkamis_pvm": 2021-01-01T00:00:00.000Z,
    "hallinnointijarjestelma": "VARDA",
    "id": "13",
    "lahdejarjestelma": "57",
    "muutos_pvm": "2022-02-09T11:29:00.249Z",
    "nimi": "Toimipaikka 3",
    "organisaatio_oid": "",
    "paattymis_pvm": null,
    "tunniste": "tunniste3",
    "url": "localhost/toimipaikka/13",
    "vakajarjestaja": "localhost/organisaatio/2.2.2.3",
    "vakajarjestaja_oid": "2.2.2.3",
  },
]
`;

exports[`should import, transform and transfer varda data 1`] = `
Array [
  Object {
    "etunimet": "Christian",
    "henkilo": "localhost/henkilo/21",
    "henkilo_id": "21",
    "henkilo_oid": "1.1.1.1",
    "id": "11",
    "kutsumanimi": "Christian",
    "lahdejarjestelma": "93",
    "lapsi": Array [
      "localhost/lapsi/11",
      "localhost/lapsi/13",
    ],
    "muutos_pvm": "2022-02-09T11:29:00.249Z",
    "oma_organisaatio": null,
    "oma_organisaatio_nimi": null,
    "oma_organisaatio_oid": null,
    "paos_kytkin": false,
    "paos_organisaatio": null,
    "paos_organisaatio_nimi": null,
    "paos_organisaatio_oid": null,
    "sukunimi": "Child",
    "syntyma_pvm": 2019-06-01T00:00:00.000Z,
    "tunniste": "tunniste1",
    "turvakielto": false,
    "tyontekija": Array [],
    "url": "localhost/lapsi/11",
    "vakatoimija": "localhost/organisaatio/2.2.2.1",
    "vakatoimija_oid": "2.2.2.1",
    "varhaiskasvatuspaatokset_top": Array [
      "localhost/paatos/1",
      "localhost/paatos/2",
    ],
  },
  Object {
    "etunimet": "Christian",
    "henkilo": "localhost/henkilo/21",
    "henkilo_id": "21",
    "henkilo_oid": "1.1.1.1",
    "id": "13",
    "kutsumanimi": "Christian",
    "lahdejarjestelma": "57",
    "lapsi": Array [
      "localhost/lapsi/11",
      "localhost/lapsi/13",
    ],
    "muutos_pvm": "2022-02-09T11:29:47.317Z",
    "oma_organisaatio": null,
    "oma_organisaatio_nimi": null,
    "oma_organisaatio_oid": null,
    "paos_kytkin": false,
    "paos_organisaatio": null,
    "paos_organisaatio_nimi": null,
    "paos_organisaatio_oid": null,
    "sukunimi": "Child",
    "syntyma_pvm": 2019-06-01T00:00:00.000Z,
    "tunniste": "tunniste2",
    "turvakielto": false,
    "tyontekija": Array [],
    "url": "localhost/lapsi/13",
    "vakatoimija": "localhost/organisaatio/2.2.2.3",
    "vakatoimija_oid": "2.2.2.3",
    "varhaiskasvatuspaatokset_top": Array [],
  },
  Object {
    "etunimet": "Camilla",
    "henkilo": "localhost/henkilo/22",
    "henkilo_id": "22",
    "henkilo_oid": "1.1.1.2",
    "id": "12",
    "kutsumanimi": "Camilla",
    "lahdejarjestelma": "45",
    "lapsi": Array [
      "localhost/lapsi/12",
    ],
    "muutos_pvm": "2022-02-09T11:29:00.249Z",
    "oma_organisaatio": "localhost/organisaatio/2.2.2.4",
    "oma_organisaatio_nimi": "Kaupunki 1",
    "oma_organisaatio_oid": "2.2.2.4",
    "paos_kytkin": true,
    "paos_organisaatio": "localhost/organisaatio/2.2.2.2",
    "paos_organisaatio_nimi": "Perhepäivähoitaja 1",
    "paos_organisaatio_oid": "2.2.2.2",
    "sukunimi": "Child",
    "syntyma_pvm": 2018-11-25T00:00:00.000Z,
    "tunniste": null,
    "turvakielto": true,
    "tyontekija": Array [
      "localhost/tyontekija/1",
    ],
    "url": "localhost/lapsi/12",
    "vakatoimija": null,
    "vakatoimija_oid": null,
    "varhaiskasvatuspaatokset_top": Array [
      "localhost/paatos/3",
    ],
  },
]
`;

exports[`should import, transform and transfer varda data 2`] = `
Array [
  Object {
    "evaka_person_id": Any<String>,
    "organizer_oid": "2.2.2.1",
    "row_id": Any<String>,
    "varda_child_id": "11",
    "varda_person_id": "21",
    "varda_person_oid": "1.1.1.1",
  },
  Object {
    "evaka_person_id": Any<String>,
    "organizer_oid": "2.2.2.2",
    "row_id": Any<String>,
    "varda_child_id": "12",
    "varda_person_id": "22",
    "varda_person_oid": "1.1.1.2",
  },
  Object {
    "evaka_person_id": Any<String>,
    "organizer_oid": "2.2.2.3",
    "row_id": Any<String>,
    "varda_child_id": "13",
    "varda_person_id": "21",
    "varda_person_oid": "1.1.1.1",
  },
]
`;

exports[`should import, transform and transfer varda data 3`] = `
Array [
  Object {
    "created_at": Any<Date>,
    "evaka_person_id": Any<String>,
    "modified_at": Any<Date>,
    "organizer_oid": "2.2.2.1",
    "uploaded_at": null,
    "varda_child_id": "11",
    "varda_person_id": "21",
    "varda_person_oid": "1.1.1.1",
  },
  Object {
    "created_at": Any<Date>,
    "evaka_person_id": Any<String>,
    "modified_at": Any<Date>,
    "organizer_oid": "2.2.2.2",
    "uploaded_at": null,
    "varda_child_id": "12",
    "varda_person_id": "22",
    "varda_person_oid": "1.1.1.2",
  },
  Object {
    "created_at": Any<Date>,
    "evaka_person_id": Any<String>,
    "modified_at": Any<Date>,
    "organizer_oid": "2.2.2.3",
    "uploaded_at": null,
    "varda_child_id": "13",
    "varda_person_id": "21",
    "varda_person_oid": "1.1.1.1",
  },
]
`;
